NuRadioReco/modules/io/coreas/coreas.py

```diff
@@ -1,58 +1,533 @@
+    Therefore the zenith angle is the same, but the azimuthal angle has to be shifted by 180 + 90 degrees.
+    The north has to be shifted by 90 degrees plus difference between geomagetic and magnetic north.
+
+
+    Parameters
+    ----------
+    corsika : hdf5 file object
+        the open hdf5 file object of the corsika hdf5 file
+    declination : float
+        declination of the magnetic field, in internal units
+
+    Returns
+    -------
+    zenith : float
+        zenith angle
+    azimuth : float
+        azimuth angle
+    magnetic_field_vector : np.ndarray
+        magnetic field vector
     """
-    zenith = np.deg2rad(corsika['inputs'].attrs["THETAP"][0])
-    azimuth = hp.get_normalized_angle(3 * np.pi / 2. + np.deg2rad(corsika['inputs'].attrs["PHIP"][0]))
+    # TODO: verify sign of declination
+
+    zenith = corsika['inputs'].attrs["THETAP"][0] * units.deg
+    azimuth = hp.get_normalized_angle(
+        3 * np.pi / 2. + np.deg2rad(corsika['inputs'].attrs["PHIP"][0]) + declination / units.rad
+    ) * units.rad
 
-    Bx, Bz = corsika['inputs'].attrs["MAGNET"]
-    B_inclination = np.arctan2(Bz, Bx)
+    # in CORSIKA convention, the first component points North (y in NRR) and the second component points down (minus z)
+    By, minBz = corsika['inputs'].attrs["MAGNET"]
+    B_inclination = np.arctan2(minBz, By)  # angle from y-axis towards negative z-axis
 
-    B_strength = (Bx ** 2 + Bz ** 2) ** 0.5 * units.micro * units.tesla
+    B_strength = np.sqrt(By ** 2 + minBz ** 2) * units.micro * units.tesla
 
-    # in local coordinates north is + 90 deg
-    magnetic_field_vector = B_strength * hp.spherical_to_cartesian(np.pi * 0.5 + B_inclination, 0 + np.pi * 0.5)
+    # zenith of the magnetic field vector is 90 deg + inclination, as inclination proceeds downwards from horizontal
+    # azimuth of the magnetic field vector is 90 deg - declination, as declination proceeds clockwise from North
+    magnetic_field_vector = B_strength * hp.spherical_to_cartesian(
+        np.pi / 2 + B_inclination, np.pi / 2 - declination / units.rad
+    )
 
     return zenith, azimuth, magnetic_field_vector
 
 
+def get_geomagnetic_angle(zenith, azimuth, magnetic_field_vector):
+    """
+    Calculates the angle between the geomagnetic field and the shower axis defined by `zenith` and `azimuth`.
+
+    Parameters
+    ----------
+    zenith : float
+        zenith angle (in internal units)
+    azimuth : float
+        azimuth angle (in internal units)
+    magnetic_field_vector : np.ndarray
+        The magnetic field vector in the NRR coordinate system (x points East, y points North, z points up)
+
+    Returns
+    -------
+    geomagnetic_angle : float
+        geomagnetic angle
+    """
+    shower_axis_vector = hp.spherical_to_cartesian(zenith / units.rad, azimuth / units.rad)
+    geomagnetic_angle = hp.get_angle(magnetic_field_vector, shower_axis_vector) * units.rad
```

Comment:
I understand you want to be "super safe" here. But I think the idea of the unit system of NuRadio is that once we read external data in we do not have to use units ever again until we start plotting or exporting data again. This also means we will _never_ change the default units.